pkg/grpc/simple_stream_forwarder.go

```diff
@@ -0,0 +1,88 @@
+package grpc
+
+import (
+	"io"
+
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+// NewSimpleStreamForwarder creates a grpc.StreamHandler that forwards gRPC
+// calls to a grpc.ClientConnInterface backend.
+func NewSimpleStreamForwarder(client grpc.ClientConnInterface) grpc.StreamHandler {
+	forwarder := &simpleStreamForwarder{
+		backend: client,
+	}
+	return forwarder.HandleStream
+}
+
+type simpleStreamForwarder struct {
+	backend grpc.ClientConnInterface
+}
+
+// HandleStream creates a new stream to the backend. Requests from
+// incomingStream are forwarded to the backend stream and responses from the
+// backend stream are sent back in the incomingStream.
+func (s *simpleStreamForwarder) HandleStream(srv any, incomingStream grpc.ServerStream) error {
+	method := MustStreamMethodFromContext(incomingStream.Context())
+	desc := grpc.StreamDesc{
+		// According to grpc.StreamDesc documentation, StreamName and Handler
+		// are only used when registering handlers on a server.
+		StreamName: "",
+		Handler:    nil,
+		// Streaming behaviour is wanted, single message is treated the same on
+		// transport level, the application just closes the stream after the
+		// first message.
+		ServerStreams: true,
+		ClientStreams: true,
+	}
+	group, groupCtx := errgroup.WithContext(incomingStream.Context())
+	outgoingStream, err := s.backend.NewStream(groupCtx, &desc, method)
+	if err != nil {
+		return err
+	}
+	go func() {
+		defer outgoingStream.CloseSend()
```

Comment:
My gut feeling tells me that we should only call `outgoingStream.CloseSend()` if we truly see `incomingStream.RecvMsg()` return `io.EOF`. Otherwise we may initiate a graceful shutdown of the outgoing stream, even if the incoming stream failed.